engine/src/ui/canvas/draw_list.cpp

```diff
@@ -0,0 +1,91 @@
+#include <cubos/core/log.hpp>
+
+#include <cubos/engine/ui/canvas/draw_command.hpp>
+#include <cubos/engine/ui/canvas/draw_list.hpp>
+#include <cubos/engine/ui/canvas/draw_type.hpp>
+
+cubos::engine::UIDrawListEntryBuilder::UIDrawListEntryBuilder(const UIDrawType& type, core::gl::VertexArray vertexArray,
+                                                              size_t vertexOffset, size_t vertexCount)
+    : mEntry(type)
+{
+    UIDrawCommand command{vertexArray, vertexOffset, vertexCount, {}, {}};
+    mEntry.command = command;
+}
+
+cubos::engine::UIDrawListEntryBuilder& cubos::engine::UIDrawListEntryBuilder::setTexture(
+    int texIndex, cubos::core::gl::Texture2D texture, cubos::core::gl::Sampler sampler)
+{
+    if (texIndex >= UIDrawType::MaxTextures)
+    {
+        CUBOS_WARN("Invalid texture index, texture will be dropped");
+        return *this;
+    }
+    mEntry.command.samplers[texIndex] = sampler;
+    mEntry.command.textures[texIndex] = texture;
+    return *this;
+}
+
+cubos::engine::UIDrawListEntry cubos::engine::UIDrawListEntryBuilder::entry() const
+{
+    return mEntry;
+}
+
+cubos::engine::UIDrawList::UIDrawList()
+{
+    mData = operator new(sizeof(char) * mDataCapacity);
+}
+
+cubos::engine::UIDrawList::UIDrawList(UIDrawList&& other) noexcept
+    : mEntries(std::move(other.mEntries))
+    , mData(other.mData)
+{
+    other.mData = nullptr;
+}
+
+cubos::engine::UIDrawList::UIDrawList(const UIDrawList& other) noexcept
+    : mEntries(other.mEntries)
+{
+    mData = operator new(sizeof(char) * other.mDataSize);
+    std::memcpy(mData, other.mData, sizeof(char) * other.mDataSize);
+}
+
+cubos::engine::UIDrawList::~UIDrawList()
+{
+    operator delete(mData);
+}
+
+void cubos::engine::UIDrawList::push(UIDrawListEntry entry, const void* data)
+{
+    entry.offset = mDataSize;
+    mEntries.push_back(entry);
+    if (mDataSize + entry.type.perElementSize > mDataCapacity)
+    {
+        mDataCapacity += entry.type.perElementSize;
```

Comment:
```suggestion
        mDataCapacity += entry.type.perElementSize;
        mDataCapacity *= 2;
```
Better to also grow it exponentially to avoid having to make many small allocations.